### PR TITLE
Interactive API Provider

### PR DESCRIPTION
**Pull request by kswenson**
[[#172312303]](https://www.pivotaltracker.com/story/show/172312303)

This PR turned into a bit of an odyssey. I was trying to keep the commits coherent so that it could be reviewed in logical chunks by commit but in the end I'm not sure it worked out well enough for that to be any better than just reviewing everything in one fell swoop. Probably wise to pack a lunch. 😉 

In broad strokes, the work can be broken down into:
- **InteractiveApiProvider** (new)
    - saves/restores state via LARA interactive API (getInteractiveState/setInteractiveState)
    - loads initial state from external document if `documentId` url parameter is provided
    - if `runRemoteEndpoint` is provided passes it on to client via `logLaraData()` callback
    - engaged via `interactiveApi` url parameter
- **Sharing dialog** (improved)
    - rewritten in modern React functional/class components
    - full suite of jest tests
    - additional tab for Interactive API sharing (enabled via `enableInteractiveApiSharing` CFM option)
- **New examples** demonstrating the above
- **Miscellaneous/unrelated typing/testing improvements** that occurred to me along the way
    - LaraProvider, UrlProvider, createReactFactory, createReactClassFactory, screenfull, etc.

**[CFM-only Example](https://activity-player.concord.org/branch/master/?activity=https%3A%2F%2Fauthoring.staging.concord.org%2Fapi%2Fv1%2Factivities%2F21084.json&page=page_309186)**: Activity Player preview of activity on learn.staging which embeds an IframeInteractive which points to a CFM example hosted at the branch URL of this PR which autosaves through the Activity Player via the interactive API (and therefore ultimately to Firestore where other answers get saved).

**[CODAP/CFM Example](https://activity-player.concord.org/branch/master/?activity=https%3A%2F%2Fauthoring.staging.concord.org%2Fapi%2Fv1%2Factivities%2F21081.json&page=page_309177)**: Activity player preview of activity on learn.staging which embeds an IframeInteractive which points to a branch build of CODAP ([PR #391](https://github.com/concord-consortium/codap/pull/391)) built with an updated version of the CFM that includes the contents of this PR. This allows CODAP document contents (up to 1MB due to Firestore limitations) to be saved via the interactive API.

Once this PR is merged:
- the CFM-only example URL should be updated to point to master rather than the PR branch
- the CFM version number should probably be incremented to the next minor version (1.3.0?)
- the corresponding [CODAP PR](https://github.com/concord-consortium/codap/pull/391) should be updated to use the CFM master version
- once the CODAP PR is merged, the CODAP/CFM example url can be updated to point to CODAP production

### CODE REVIEW

File: src/code/views/share-dialog-tabs-view.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+          switch (tabSelected) {
+            case 'link':
+              return <LinkTabContents url={linkUrl} onCopyClick={onCopyClick} />
+            case 'embed':
+              return <EmbedTabContents url={embedUrl} onCopyClick={onCopyClick} />
+            case 'lara': {
+              const { linkUrl: url, ...laraOthers } = lara
+              return <LaraApiTabContents mode='lara' linkUrl={url} onCopyClick={onCopyClick}
+                                        {...others} {...laraOthers} />
+            }
```
Comment (scytacki) [inline src/code/views/share-dialog-tabs-view.tsx:193]: Seems more consistent to use `{..}` in each case if you have to use it here.

**Review by scytacki — APPROVED** [review]
Looks good to me. I just found a few things to comment on, but they don't need to hold up the merge.
I didn't get a chance to test the links, but I figure I'll see them at the upcoming demo, otherwise I'll check them out soon.

File: src/code/providers/interactive-api-provider.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
+import queryString from 'query-string'
+import React from 'react'
+import { CFMInteractiveApiProviderOptions, CFMLaraProviderLogData } from '../app-options'
+import { CloudFileManagerClient } from '../client'
+import {
+  cloudContentFactory, CloudMetadata, ECapabilities, ProviderInterface,
+  ProviderLoadCallback, ProviderOpenCallback, ProviderSaveCallback
+}  from './provider-interface'
+import {
+  getInitInteractiveMessage, getInteractiveState, IRuntimeInitInteractive, setInteractiveState
```
[the remark this answers is not in the document]
Comment (kswenson) [reply src/code/providers/interactive-api-provider.ts:10]: I'm inclined to leave this for another story.

File: src/code/views/share-dialog-tabs-view.tsx (hunk 1 above)
Comment (kswenson) [reply src/code/views/share-dialog-tabs-view.tsx:193]: The `link` and `embed` tabs only need two props. All the others are for the use of the `lara` and `api` tabs.

**Pull request merged**